### PR TITLE
Adds Bowe-Hopwood hash

### PR DESCRIPTION
**Pull request by kobigurk**
* Adds 3-bit lookup with conditional negation in Groups
* Adds Bowe-Hopwood implementation according to the Zcash spec
* Adds add_bool_as_coeff and three_bit_cond_neg in Fields

### CODE REVIEW

File: crypto-primitives/src/crh/bowe_hopwood/mod.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    }
+}
+
+impl<G: Group, W: PedersenWindow> FixedLengthCRH for BoweHopwoodPedersenCRH<G, W> {
+    const INPUT_SIZE_BITS: usize = PedersenCRH::<G, W>::INPUT_SIZE_BITS;
+    type Output = G;
+    type Parameters = BoweHopwoodPedersenParameters<G>;
+
+    fn setup<R: Rng>(rng: &mut R) -> Result<Self::Parameters, Error> {
+        fn calculate_num_chunks_in_segment<G: Group>() -> usize {
```
Comment (Pratyush) [inline crypto-primitives/src/crh/bowe_hopwood/mod.rs]:
```suggestion
        fn calculate_num_chunks_in_segment<F: PrimeField>() -> usize {
```

File: crypto-primitives/src/crh/bowe_hopwood/mod.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            let upper_limit = G::ScalarField::modulus_minus_one_div_two();
```
Comment (Pratyush) [inline crypto-primitives/src/crh/bowe_hopwood/mod.rs]:
```suggestion
            let upper_limit = F::modulus_minus_one_div_two();
```

File: crypto-primitives/src/crh/bowe_hopwood/mod.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            let mut c = 0;
+            let mut range = <G::ScalarField as PrimeField>::BigInt::from(2_u64);
```
Comment (Pratyush) [inline crypto-primitives/src/crh/bowe_hopwood/mod.rs]:
```suggestion
            let mut range = F::BigInt::from(2_u64);
```

File: crypto-primitives/src/crh/bowe_hopwood/mod.rs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            while range < upper_limit {
+                range.muln(4);
+                c += 1;
+            }
+
+            c
+        }
+
+        let maximum_num_chunks_in_segment = calculate_num_chunks_in_segment::<G>();
```
Comment (Pratyush) [inline crypto-primitives/src/crh/bowe_hopwood/mod.rs]:
```suggestion
        let maximum_num_chunks_in_segment = calculate_num_chunks_in_segment::<G::ScalarField>();
```

File: crypto-primitives/src/crh/bowe_hopwood/mod.rs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        ));
+        let generators = Self::create_generators(rng);
+        end_timer!(time);
+        Ok(Self::Parameters {
+            generators
+        })
+    }
+
+    fn evaluate(parameters: &Self::Parameters, input: &[u8]) -> Result<Self::Output, Error> {
+        let eval_time = start_timer!(|| "PedersenCRH::Eval");
```
Comment (Pratyush) [inline crypto-primitives/src/crh/bowe_hopwood/mod.rs]:
```suggestion
        let eval_time = start_timer!(|| "BoweHopwoodPedersenCRH::Eval");
```

File: r1cs-std/src/fields/mod.rs [hunk 6]
```diff
@@ -38,6 +39,8 @@ pub trait FieldGadget<F: Field, ConstraintF: Field>:
 
     fn one<CS: ConstraintSystem<ConstraintF>>(_: CS) -> Result<Self, SynthesisError>;
 
+    fn add_bool_with_coeff<CS: ConstraintSystem<ConstraintF>>(&self, _: CS, _: &Boolean, _: F) -> Result<Self, SynthesisError>;
```
Comment (Pratyush) [inline r1cs-std/src/fields/mod.rs]: Maybe rename this to `conditionally_add_constant`? Currently it feels like you're adding the bool value to `self`.

File: r1cs-std/src/groups/curves/twisted_edwards/mod.rs [hunk 7]
```diff
@@ -15,6 +15,206 @@ pub mod jubjub;
...
+        pub fn new(x: F, y: F) -> Self {
+            Self {
+                x,
+                y,
+                _params: PhantomData,
+                _engine: PhantomData,
+            }
+        }
+
+        pub fn params() -> MontgomeryParams<P> {
```
Comment (Pratyush) [inline r1cs-std/src/groups/curves/twisted_edwards/mod.rs]: Longer term it could be nice to have a `MontgomeryParameters` trait (like the `TEModelParameters` trait), so that the `A` and `B` values generated here can be fixed constants.

File: r1cs-std/src/groups/curves/twisted_edwards/mod.rs [hunk 8]
```diff
@@ -757,6 +957,122 @@ mod projective_impl {
...
+                    }
+                    Some(ref mut edwards_result) => {
+                        *edwards_result = GroupGadget::<TEAffine<P>, ConstraintF>::add(
+                            &segment_result,
+                            cs.ns(|| "edwards addition of leftover"),
+                            edwards_result,
+                        )?;
+                    }
+                }
+            }
```
Comment (Pratyush) [inline r1cs-std/src/groups/curves/twisted_edwards/mod.rs]: These pieces of code are almost exactly the same, so it might be good to extract it into a macro or function/closure that is local to `precomputed_base_scalar_mul_3_bit_with_conditional_negation`?

File: r1cs-std/src/groups/mod.rs [hunk 9]
```diff
@@ -111,6 +111,21 @@ pub trait GroupGadget<G: Group, ConstraintF: Field>:
         Ok(())
     }
 
+    fn precomputed_base_scalar_mul_3_bit_with_conditional_negation<'a, CS, T, I, B>(
```
Comment (Pratyush) [inline r1cs-std/src/groups/mod.rs]: From what I understand, the only difference between `precomputed_base_scalar_mul_3_bit_with_conditional_negation` and `precomputed_base_multiscalar_mul` is that the former requires an additional `segment_size` argument. Is there a way to avoid duplication in the interface itself?

Couple of options:
1) Have just one generic `precomputed_base_multiscalar_mul` whose signature is the same as `precomputed_base_scalar_mul_3_bit_with_conditional_negation`, but whose default implementation is the same as the current `precomputed_base_multiscalar_mul`. Then, for `TEGadget`, we just use the current implementation of `precomputed_base_scalar_mul_3_bit_with_conditional_negation` as the impl.
2) Investigate whether we actually need the `segment_size` argument, or whether it can be a fixed constant derivable from things like the scalar field size? I understand that, as used in the BHP hash, `segment_size` is just the window-size; is this value expected to differ across instantiations, or will it remain the same?

**Review by Pratyush — CHANGES_REQUESTED** [review]
Thanks for the PR! Left a couple of quick typo fix suggestions, as well as a couple of questions regarding some code structure/naming.

File: r1cs-std/src/groups/mod.rs (hunk 9 above)
Comment (kobigurk) [reply r1cs-std/src/groups/mod.rs]: That might be problematic - precomputed_base_scalar_mul_3_bit_with_conditional_negation doesn't have the same goal as precomputed_base_multiscalar_mul - it's more efficient, though it maps ([b1, b2, b3], G) to [-4,...,-1,1,...4]*G rather than [0,...,7]*G as precomputed_base_multiscalar_mul does.

I think we might be able to get rid of segment_size though.

File: r1cs-std/src/fields/mod.rs (hunk 6 above)
Comment (kobigurk) [reply r1cs-std/src/fields/mod.rs]: Sounds great.

File: r1cs-std/src/groups/curves/twisted_edwards/mod.rs (hunk 7 above)
Comment (kobigurk) [reply r1cs-std/src/groups/curves/twisted_edwards/mod.rs]: Good idea, I added it.
It does require every Edwards curve to provide the Montgomery parameters.

File: r1cs-std/src/groups/mod.rs (hunk 9 above)
Comment (kobigurk) [reply r1cs-std/src/groups/mod.rs]: Not 100% sure how to cleanly get rid of segment size yet. Let me know if you got ideas.

File: r1cs-std/src/groups/curves/twisted_edwards/mod.rs (hunk 8 above)
Comment (kobigurk) [reply r1cs-std/src/groups/curves/twisted_edwards/mod.rs]: Take a look at the latest commit.

File: r1cs-std/src/groups/mod.rs (hunk 9 above)
Comment (kobigurk) [reply r1cs-std/src/groups/mod.rs]: The reason it's not direct is because it's not directly a property of the group when to move on to the next segment. The maximum is a property though. The question is whether to always use the maximum.

Comment (Pratyush) [reply r1cs-std/src/groups/mod.rs]: Re: merging the two, then maybe it might be worthwhile to rename the method to something different? To avoid confusion with normal multi-scalar multiplication. Do you have any ideas? (`precomputed_base_scalar_mul_3_bit_with_conditional_negation` is also a bit wordy)

If we rename it, then maybe the segment_size question is also moot, if we make it clearer what it means?. Another approach would be to potentially derive `segment_size` from the input length + group size? Do you have any references I could look at that talk about the derivation of segment_size?

Comment (kobigurk) [reply r1cs-std/src/groups/mod.rs]: That's a good idea - let's rename it. I'll think of something better.

The derivation of the maximum segment size is both implemented in https://github.com/celo-org/zexe/blob/052e66bb6153383f164aba8a3847f0e13bb2e903/crypto-primitives/src/crh/bowe_hopwood/mod.rs#L57 and described in the Zcash spec (https://github.com/zcash/zips/blob/master/protocol/protocol.pdf) in page 56.

Basically, it's the amount of generators you can use in this method and still not pass the maximum scalar.

If we always use the maximum one, then no doubt we can derive it. I think we might want to do that, there shouldn't be any real advantage to doing otherwise.

**Comment by Pratyush** [comment]
The PR looks good, modulo the question regarding `precomputed_base_scalar_mul_3_bit_with_conditional_negation`, and the following point:

Instead of adding the `P: MontgomeryModelParameters` bound everywhere, do you think it would be better to add an associated type to `TEModelParameters` as follows:

```rust
pub trait TEModelParameters: ModelParameters {
	type MontModelParameters: MontgomeryParameters<BaseField == Self::BaseField> + From<Self>;
    // Rest here ...
}
```
and similarly in `MontgomeryModelParameters`:

```rust
pub trait MontgomeryModelParameters: ModelParameters {
	type TEModelParameters: TEModelParameters<BaseField == Self::BaseField> + From<Self>;
    // Rest here ...
}
```
This way one can just refer to `P::MontModelParameters` in the relevant places, while reducing the duplicated trait bounds. It would also allow us to rewrite `<P as TEModelParameters>::COEFF_A` to `P::COEFF_A`, which is slightly cleaner.

Let me know what you think; we can also land as-is and I can refactor as needed

**Comment by kobigurk** [comment]
> The PR looks good, modulo the question regarding `precomputed_base_scalar_mul_3_bit_with_conditional_negation`, and the following point:
> 
> Instead of adding the `P: MontgomeryModelParameters` bound everywhere, do you think it would be better to add an associated type to `TEModelParameters` as follows:
> 
> ```rust
> pub trait TEModelParameters: ModelParameters {
> 	type MontModelParameters: MontgomeryParameters<BaseField == Self::BaseField> + From<Self>;
>     // Rest here ...
> }
> ```
> 
> and similarly in `MontgomeryModelParameters`:
> 
> ```rust
> pub trait MontgomeryModelParameters: ModelParameters {
> 	type TEModelParameters: TEModelParameters<BaseField == Self::BaseField> + From<Self>;
>     // Rest here ...
> }
> ```
> 
> This way one can just refer to `P::MontModelParameters` in the relevant places, while reducing the duplicated trait bounds. It would also allow us to rewrite `<P as TEModelParameters>::COEFF_A` to `P::COEFF_A`, which is slightly cleaner.
> 
> Let me know what you think; we can also land as-is and I can refactor as needed

That's a great idea! I think worth doing before merging.

File: r1cs-std/src/groups/mod.rs (hunk 9 above)
Comment (Pratyush) [reply r1cs-std/src/groups/mod.rs]: I'll leave the decision regarding `segment_size` up to you! Re: name suggestions, maybe something based on `signed_digits`? Eg:

* `precomputed_base_signed_digit_scalar_mul`, with a comment explaining that it works only for 3-bit signed digits?
* `precomputed_base_3_bit_signed_digit_scalar_mul`

**Comment by kobigurk** [comment]
Hey @Pratyush, 

I changed the structure of the code a bit - the responsibility on chunking the segments appropriately before getting to the `precomputed_base_3_bit_signed_digit_scalar_mul ` is now passed on to the caller - so there's no segment size needed anymore. The code looks cleaner overall, IMO.

Also made the change with the associated type.

File: algebra/src/curves/edwards_bls12/mod.rs [hunk 10]
```diff
@@ -63,6 +65,25 @@ impl TEModelParameters for EdwardsParameters {
...
 
+impl MontgomeryModelParameters for EdwardsParameters {
+    /// COEFF_A = -1
+    const COEFF_A: Fq = field_new!(Fq, BigInteger256([
+        13800168384327121454u64,
+        6841573379969807446u64,
+        12529593083398462246u64,
+        853978956621483129u64,
+    ]));
+    /// COEFF_B = -1
```
Comment (Pratyush) [inline algebra/src/curves/edwards_bls12/mod.rs]: Is this comment accurate?

File: algebra/src/curves/edwards_sw6/mod.rs [hunk 11]
```diff
@@ -62,13 +62,38 @@ impl TEModelParameters for EdwardsParameters {
...
+    /// COEFF_A = -1
+    const COEFF_A: Fq = field_new!(Fq, BigInteger([
+        7594254284108454966u64,
+        14287343397973578077u64,
+        6490358977072726023u64,
+        8023375322051995268u64,
+        8242802613686040715u64,
+        100541941146122331u64,
+    ]));
+    /// COEFF_B = -1
```
Comment (Pratyush) [inline algebra/src/curves/edwards_sw6/mod.rs]: Is the COEFF_B == -1?

File: algebra/src/curves/jubjub/mod.rs [hunk 12]
```diff
@@ -89,13 +89,34 @@ impl TEModelParameters for JubJubParameters {
...
 
+impl MontgomeryModelParameters for JubJubParameters {
+    /// COEFF_A = -1
+    const COEFF_A: Fq = field_new!(Fq, BigInteger256([
+        388496971701930u64,
+        6855257088226130262u64,
+        553476580979119549u64,
+        6516741293351590684u64,
+    ]));
+    /// COEFF_B = -1
```
Comment (Pratyush) [inline algebra/src/curves/jubjub/mod.rs]: Same as the other COEFF_B comments

File: crypto-primitives/src/crh/bowe_hopwood/constraints.rs [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+{
+    type OutputGadget = GG;
+    type ParametersGadget = BoweHopwoodPedersenCRHGadgetParameters<G, W, ConstraintF, GG>;
+
+    fn check_evaluation_gadget<CS: ConstraintSystem<ConstraintF>>(
+        cs: CS,
+        parameters: &Self::ParametersGadget,
+        input: &[UInt8],
+    ) -> Result<Self::OutputGadget, SynthesisError> {
+        // Pad the input if it is not the current length.
```
Comment (Pratyush) [inline crypto-primitives/src/crh/bowe_hopwood/constraints.rs:58]:
```suggestion
        // Pad the input if it is not the correct length.
```

File: crypto-primitives/src/crh/bowe_hopwood/mod.rs [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+            W::WINDOW_SIZE,
+            W::NUM_WINDOWS,
+            CHUNK_SIZE,
+        );
+        for generators in parameters.generators.iter() {
+            assert_eq!(generators.len(), W::WINDOW_SIZE);
+        }
+        assert_eq!(CHUNK_SIZE, 3);
+
+        // Compute sum of h_i^{m_i} for all i.
```
Comment (Pratyush) [inline crypto-primitives/src/crh/bowe_hopwood/mod.rs]: Maybe update the comment to reflect the new mapping

**Review by Pratyush — CHANGES_REQUESTED** [review]

**Comment by Pratyush** [comment]
@kobigurk I left a couple of small comments, we can merge once those are addressed. Would you want to squash some of the commits? If not that's fine too.

File: algebra/src/curves/edwards_bls12/mod.rs (hunk 10 above)
Comment (kobigurk) [reply algebra/src/curves/edwards_bls12/mod.rs]: This was a placeholder 🤦‍♂

File: algebra/src/curves/edwards_sw6/mod.rs (hunk 11 above)
Comment (kobigurk) [reply algebra/src/curves/edwards_sw6/mod.rs]: Another placeholder.

**Comment by kobigurk** [comment]
> @kobigurk I left a couple of small comments, we can merge once those are addressed. Would you want to squash some of the commits? If not that's fine too.

Alright, fixed those comments (oops), and squashed.

**Comment by Pratyush** [comment]
Awesome, thanks so much for this contribution and for your patience with my reviews!
